### PR TITLE
Search Project Setup

### PR DESCRIPTION
**Pull request by vincue**

### CODE REVIEW

File: search-api/config.py [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+
+# Application Directory
+
+import os
+BASE_DIR = os.path.abspath(os.path.dirname(__file__))
+
+# # Example DB Configuration
+SQLALCHEMY_DATABASE_URI = f"sqlite:////{BASE_DIR}/db/db.sqlite3"
+
+# SQLALCHEMY_DATABASE_URI = ''
```
Comment (kshitij-cuebook) [inline search-api/config.py:13]: We can remove the commented code before merging.

File: search-api/entrypoint.sh [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+	echo production 
+else
+	pip install ipython
+	pip install -r requirements.txt
+	chmod -R 777 /code/
+	chown -R www-data:www-data /code/
+fi
+export FLASK_APP=search
+# flask db init
+flask db migrate
```
Comment (kshitij-cuebook) [inline search-api/entrypoint.sh:14]: migrate should not be a part of entrypoint - it creates new migrations files, which will be created by the developer. We can have upgrade.

File: search-api/migrations/versions/d5ad68999347_.py [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+    sa.Column('salary', sa.Numeric(), nullable=True),
+    sa.Column('references', sa.String(), nullable=True),
+    sa.PrimaryKeyConstraint('id')
+    )
+    op.create_table('global_dimension_value',
+    sa.Column('id', sa.Integer(), nullable=False),
+    sa.Column('datasetId', sa.Integer(), nullable=True),
+    sa.PrimaryKeyConstraint('id')
+    )
+    op.create_table('otherthing',
```
Comment (kshitij-cuebook) [inline search-api/migrations/versions/d5ad68999347_.py:34]: I think these are just experimental tables, we should remove these before committing.

File: search-api/search/__init__.py [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+# Configurations
+app.config.from_object('config')
+
+
+# Some more example SQLAlchemy config
+# Define the database object which is imported by modules and controllers
+db = SQLAlchemy(app)
+
+
+# # HTTP error handling
```
Comment (kshitij-cuebook) [inline search-api/search/__init__.py:21]: Let's remove the commented code.

File: search-api/search/__init__.py [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+#     return render_template('404.html'), 404
+
+# # Home page view
+# @app.route('/')
+# def home():
+#     return render_template('home.html')
+
+
+# Import modules here
+from .cardTemplates import views, models
```
Comment (kshitij-cuebook) [inline search-api/search/__init__.py:33]: We should take the imports tot he top

File: search-api/search/cardTemplates/models.py [hunk 6]
```diff
@@ -0,0 +1,15 @@
+from search import db
+
+class Something(db.Model):
```
Comment (kshitij-cuebook) [inline search-api/search/cardTemplates/models.py:3]: We need to remove these before merging (assuming Something and Otherthing are just for testing.)

**Review by kshitij-cuebook — CHANGES_REQUESTED** [review]

File: search-api/README.md [hunk 7]
```diff
@@ -0,0 +1,31 @@
+# ReadMe
+
+The code for flask application can be found in `search` folder
+Run given below codes in `search-api` directory
+
+## To setup dev environment
+```sh
+docker build -t cue-search .
```
[the remark this answers is not in the document]
Comment (vincue) [reply search-api/README.md:8]: This README is only for this microservice.

File: search-api/config.py [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+
+DEBUG = True
+
+# Application Directory
+
+import os
+BASE_DIR = os.path.abspath(os.path.dirname(__file__))
+
+# # Example DB Configuration
+SQLALCHEMY_DATABASE_URI = f"sqlite:////{BASE_DIR}/db/db.sqlite3"
```
[the remark this answers is not in the document]
Comment (vincue) [reply search-api/config.py:11]: Yeah, it'll be optional ,if you pass it in evironment variables.

[the remark this answers is not in the document]
Comment (kshitij-cuebook) [reply search-api/config.py:11]: Yes, we need to make sure the database here and in the central (core) service is the same. So, the same env variable should be used for database configuration in both the places. I am resolving this as well, assuming we will do it later.

File: search-api/migrations/versions/d5ad68999347_.py (hunk 3 above)
Comment (vincue) [reply search-api/migrations/versions/d5ad68999347_.py]: For reference purpose I've commented them.

File: search-api/search/cardTemplates/models.py (hunk 6 above)
Comment (vincue) [reply search-api/search/cardTemplates/models.py]: For reference purpose I've commented them.

File: search-api/entrypoint.sh (hunk 2 above)
Comment (kshitij-cuebook) [reply search-api/entrypoint.sh]: @vincue what about this one?

File: search-api/entrypoint.sh [hunk 9]
```diff
@@ -0,0 +1,17 @@
+#!/bin/bash
+
+if [[ $ENVIRONMENT == "PRODUCTION" ]]
+then
+	echo production 
+else
+	pip install ipython
+	pip install -r requirements.txt
+	chmod -R 777 /code/
```
Comment (kshitij-cuebook) [inline search-api/entrypoint.sh:9]: Why do we need these two lines: `chmod and chown code` ?

File: search-api/entrypoint.sh [hunk 10]
```diff
@@ -0,0 +1,17 @@
+#!/bin/bash
+
+if [[ $ENVIRONMENT == "PRODUCTION" ]]
+then
+	echo production 
+else
+	pip install ipython
+	pip install -r requirements.txt
```
Comment (kshitij-cuebook) [inline search-api/entrypoint.sh:8]: I think requirement installation should be common to both prod and dev, only some requirements will not be installed on prod. So, if that's the case, we can create two requirements.txt files - one for base requirements, and other which are just required for dev env.

File: search-api/entrypoint.sh (hunk 9 above)
Comment (vincue) [reply search-api/entrypoint.sh:9]: Giving permission for making changes in files.

File: search-api/entrypoint.sh (hunk 10 above)
Comment (vincue) [reply search-api/entrypoint.sh:8]: Yeah will separate out later. And requirements are being installed while building image, installing again so wouldn't have build everytime a requirement changes (only for dev)

File: search-api/entrypoint.sh (hunk 9 above)
Comment (vincue) [reply search-api/entrypoint.sh:9]: Will remove it from production, only needed in development.

File: search-api/entrypoint.sh (hunk 2 above)
Comment (vincue) [reply search-api/entrypoint.sh]: You're right, I've made the changes. Upgrade will do the job.

File: search-api/entrypoint.sh (hunk 9 above)
Comment (kshitij-cuebook) [reply search-api/entrypoint.sh:9]: For development, I think we will use docker-compose and we can just map the entire code volume and give it read-write permissions. This shouldn't be required. I am resolving this now, but if it is not required in future, we can remove it.

**Pull request merged**